### PR TITLE
Adding the necessary infrastructure for writing unit tests

### PR DESCRIPTION
**Pull request by darthvader88**
This change improves build system by moving the build of each sub-product into a separate subdir in build/. Also, it adds a target for building unit tests that build its own 'test_kernel' with different flags.
The idea is that unit tests will use a special "test" build of the kernel for testing it.
For the moment, no tests are present.

As a side-effect of this change, the -fleading-underscore has been removed from the CFLAGS, making the code nicer. This was necessary in order to the kernel objects to be linked with the regular executable 'unittests'.

### CODE REVIEW

File: src/kernelAsm.asm [hunk 1]
```diff
@@ -1,13 +1,13 @@
...
 
-extern _kmain
-extern _idtp
-extern _irq_handler
-extern _gdt_pointer
-
-global _gdt_load
-global _idt_load
-global _tss_flush
+extern kmain
```
Comment (ZombineDev) [inline src/kernelAsm.asm]: The code indeed lookz nicer without underscores.

File: unittests/asmMocks.c [hunk 2]
```diff
@@ -0,0 +1,55 @@
+void isr0() { }
```
Comment (ZombineDev) [inline unittests/asmMocks.c]: Yeah I had something like that in mind when we talked about making the code more testable.

File: unittests/testmain.c [hunk 3]
```diff
@@ -0,0 +1,11 @@
+#include <stdio.h>
+
+void *kmalloc(size_t);
+
+int main(int argc, char **argv) {
+
+   printf("hello from the kernel unit tests!\n");
+   printf("kmalloc(10) returns: %p\n", kmalloc(10));
+
```
Comment (ZombineDev) [inline unittests/testmain.c]: void *ptr = kmalloc(10);
assert (ptr != 0);

File: src/kernelAsm.asm (hunk 1 above)
Comment (darthvader88) [inline src/kernelAsm.asm]: Yep!

File: unittests/testmain.c (hunk 3 above)
Comment (darthvader88) [inline unittests/testmain.c]: That was not a test. That's why I did not use any assert.
My point was to try calling an independent kernel function from the tests and see that it works.

**Pull request merged**